File: Convert_an_Array_Into_a_2D_Array_With_Conditions.py
class Solution:
    def findMatrix(self, nums: List[int]) -> List[List[int]]:
        res = []
        freq = {}

        for num in nums:
            freq[num] = 1 + freq.get(num, 0)

            if freq[num] > len(res):
                res.append([])
            
            res[freq[num] - 1].append(num)
        
        return res

File: Convert_an_Array_Into_a_2D_Array_With_Conditions.cpp
class Solution {
public:
    vector<vector<int>> findMatrix(vector<int>& nums) {
        vector<vector<int>> res;
        unordered_map<int, int> freq;

        for (int num : nums) {
            freq[num] = 1 + freq[num];

            if (freq[num] > res.size()) {
                res.push_back(vector<int>());
            }

            res[freq[num] - 1].push_back(num);
        }

        return res;        
    }
};
